developer_guides/uuid/index.rst

```diff
@@ -0,0 +1,134 @@
+
+UUID allocation
+***************
+The UUID allocation of a specific component should use the version 4 in
+https://en.wikipedia.org/wiki/Universally_unique_identifier, and the
+value should be specified in the firmware component dirver, e.g. for src
+component, in sof/src/audio/src.c:
+
+.. code-block:: none
+
+   /* c1c5326d-8390-46b4-aa47-95c3beca6550 */
+   DECLARE_SOF_UUID("src", src_uuid, 0xc1c5326d, 0x8390, 0x46b4,
+                    0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50);
+
+UUID in topology
+****************
+We should provide UUID string in topology .m4 file in the exactly same
+format as commented in the firmware, e.g for host it is
+c1c5326d-8390-46b4-aa47-95c3beca6550, so in the topology .m4 file
+tools/topology/m4/src.m4:
+
+.. code-block:: none
+
+   dnl Defines the macro for SRC widget
+   define(SRC_UUID_ARRAY, UUID_ARRAY("c1c5326d-8390-46b4-aa47-95c3beca6550"))
+
+We have a .m4 macro UUID_ARRAY() will help to transfer format in #2 (36
+digits) to 16 Bytes byte array, after compiled, the corresponding ALSA
+UUID token defined here will be generated:
+
+.. code-block:: none
+
+   /* vendor tuple for uuid */
+   struct snd_soc_tplg_vendor_uuid_elem {
+           __le32 token;
+           char uuid[16];
+   } __attribute__((packed));
+
+Linux topology driver
+*********************
+The topology driver will parse the 16Bytes UUID token and sent it to the
+firmware in component creation stage, **for all components**.
+
+.. code-block:: none
+
+   /* Generic component tokens */
+   static const struct sof_topology_token comp_tokens[] = {
+           {SOF_TKN_COMP_CORE_ID,
+               SND_SOC_TPLG_TUPLE_TYPE_WORD, get_token_u32,
+               offsetof(struct sof_ipc_comp, core), 0},
+           {SOF_TKN_COMP_UUID,
+               SND_SOC_TPLG_TUPLE_TYPE_UUID, get_token_uuid,
+               offsetof(struct sof_ipc_comp, uuid), 0},
+   };
+
+.. code-block:: none
+
+   static int sof_widget_ready(struct snd_soc_component *scomp, int index,
+                               struct snd_soc_dapm_widget *w,
+                               struct snd_soc_tplg_dapm_widget *tw)
+   {
+   ...
+           ret = sof_parse_tokens(scomp, &template, comp_tokens,
+                                  ARRAY_SIZE(comp_tokens), tw->priv.array,
+                                  le32_to_cpu(tw->priv.size));
+   ...
+   }
+
+UUID arrays stored section
+**************************
+The UUID arrays are stored to SRAM as a separated section, though it
+will lead to a 1~4KB (depends on how many modules are compiled) memory
+consumption increase. e.g. in src/platform/cannonlake/cannonlake.x.in:
+
+.. code-block:: none
+
+  .static_uuid_entries : ALIGN(4)
+  {
+    *(*.static_uuids)
+  } >sof_fw :sof_fw_phdr
+
+UUID to component driver mapping
+********************************
+The firmware will use UUID byte array as the first choice to match the
+component driver, and fallback to use traditional component type if
+failed:
```

Comment:
why? we don't need to support the old kernel/new firmware model. it'd be perfectly fine to *require* uuids to be provided for each component.

Reply:
@plbossart Do we see value in being able to define topologies with component type. E.g. one could define a topology with a volume component. Without uuid, it would pick whatever implementation the DSP has (component_type==VOLUME). These topologies would work with multiple imlementations of volume.

Reply:
@kv2019i that sounds like asking for trouble? If you have multiple implementations of the volume, then you may have different parameters or configurations, so likely to have the driver provide a config that the firmware doesn't understand. I would really err on the side of strong-types with no ambiguity: topology, firmware and driver all have a common understanding of what each component on each path do and how to configure them. I think it's even more important for 3rd party, e.g. you may have a plugin version on one path and a different one on the other due to MCPS/memory/performance/acoustics requirements.

Reply:
"type" will be deprecated and removed with IPC2.0, this just give some overlap.
@keyonjie type will only be used if UUID is missing, if UUID fails then there is no match.

Reply:
@lgirdwood 'type' as a component identifier will no longer be used, but I think we need to keep a 'type' or 'class' that tells us what a component is (DAI, effect, decoder, etc). Maybe you don't pass it over IPC but it's part of the component definition.

Reply:
And second @lgirdwood why do we need a fallback mechanism? we can bump the firmware and expect the kernel to use the UUIDs.

Reply:
@plbossart agree, driver still needs some limited types for DAPM. The falback was to help with combinations of different ABIs between FW, topology and driver. Removing type from IPC ABI would be a MAJOR and it would be best to do this with IPC2.0 MAJOR.

Reply:
Yes, this is only try to give some overlap during the transition period, somebody might use old tplg files, or UUID of some component may not be specified in the tplg files, we will throw out a warning for this kind of mistake at the moment, and we can remove this fallback mechanism when we should.

Reply:
@lgirdwood @plbossart Do we agree that if we change semantics of "component_type", this is a MAJOR bump? I know we haven't committed to support new fw with old kernels, but I think we should stick to semantic versioning and I think this is a MAJOR bump. OTOH, +95% of users will update the topologies at the same time, so in practise SOF release will require a very recent kernel to use.

Reply:
if UUID is not present then "type" is used using existing IPC
if UUID is present in topology then we use new IPC message (MINOR update)
if UUID is present in topology and FW is too old we use "type" and old message (MINOR).

Reply:
We have 8 combination cases here so it is actually somewhat complicate, if we can use major bump and Force "driver ABI major == FW ABI major", it will make our life easier.


FW | tplg | driver | supported today? | Force driver ABI >= FW ABI | Use major bump && Force driver ABI major == FW ABI major | comment
-- | -- | -- | -- | -- | -- | --
old | old | old | Y | support | support |  
old | old | new | support if component type fallback support in driver with old ABI | support if component type fallback support in driver with old ABI | rejected |  
old | new | old | support, uuid in tplg doesn’t take effect | support, uuid in tplg doesn’t take effect | support, uuid in tplg doesn’t take effect |  
old | new | new | support if component type fallback support in driver with old ABI | support if component type fallback support in driver with old ABI | rejected |  
new | old | old | N, buggy. | rejected | rejected |  
new | old | new | support, need support component type fallback in both fw&driver | support, need support component type fallback in both fw&driver | support, need support component type fallback in both fw&driver |  
new | new | old | N, buggy. | rejected | rejected |  
new | new | new | Y | support | support |  

Reply:
@keyonjie Ok, that's a very thorough table.  I think you are muddying the waters w.r.t. to our agreed semantics for ABI versioning here though:
 - If driver ABI version is forced to be same-or-newer than FW, what's the point to maintain MINOR versions at all?
 - I don't get the "supported today?" column... do you mean the option that UUID is added to the IPC as a MINOR option? 
- if yes to the above, I don't understand the "N, buggy" entries. there's still the comp_type, so I don't see what the bugs would be

I'd prefer to make this a simple call between 
- a)  UUID is a minor bump (we can support old IPC interface in FW), and
- b) UUID is a major bump (we cannot support the old IPC)

a or b, but let's not start changing our versioning semantics mid-flight.

Reply:
> @keyonjie Ok, that's a very thorough table. I think you are muddying the waters w.r.t. to our agreed semantics for ABI versioning here though:
> 
> * If driver ABI version is forced to be same-or-newer than FW, what's the point to maintain MINOR versions at all?

I am not sure if we already have assumption that the driver will be newer than the FW in most of the time, from the code, by default if CONFIG_SND_SOC_SOF_STRICT_ABI_CHECKS is not specified, we will only throw a warning only if the FW is newer. This could be a risk, in this UUID case, if FW is newer and try to read the UUID bytes from an old driver IPC struct, out of bounds memory access could happen. From this angle, do you think we should do a major bump for UUID update?
```
	if (v->abi_version > SOF_ABI_VERSION) {
		if (!IS_ENABLED(CONFIG_SND_SOC_SOF_STRICT_ABI_CHECKS)) {
			dev_warn(sdev->dev, "warn: FW ABI is more recent than kernel\n");
		} else {
			dev_err(sdev->dev, "error: FW ABI is more recent than kernel\n");
			return -EINVAL;
		}
	}
```


> * I don't get the "supported today?" column... do you mean the option that UUID is added to the IPC as a MINOR option?

Yes.

> * if yes to the above, I don't understand the "N, buggy" entries. there's still the comp_type, so I don't see what the bugs would be

If we keep the comp_type in the struct sof_ipc_comp and append the uuid to the end, looks we are fine in most of the time (with only MINOR bump), but if the comp_type is not come from the driver anymore (e.g. the comp_type is moved out of the struct sof_ipc_comp) then we have to do a major bump, otherwise, it will be buggy.

> 
> I'd prefer to make this a simple call between
> 
> * a)  UUID is a minor bump (we can support old IPC interface in FW), and
> * b) UUID is a major bump (we cannot support the old IPC)
> 
> a or b, but let's not start changing our versioning semantics mid-flight.

Makes sense, if we are aligned to keep the comp_type in the struct sof_ipc_comp and append the uuid to the end at this moment, we can do only a minor bump with a, and do major bump b maybe in some day that the comp_type is moved out of the struct sof_ipc_comp .

Reply:
@keyonjie we will include **both** the type and UUID in topology until the IPC2.0 MAJOR bump. This means we are good for any combination in teh table, the fallback is always to use the type and only use UUID is supported by all FW, driver and topology.

Reply:
@lgirdwood proposed a solution to fix the ABI backward compatible issue to me, it basically put the UUID array to append to the whole component structure, and use the sof_ipc_comp.reserved member as its offset to the component struct.

Reply:
the PR is updated with this part changed, mark it 'resolved'.